crates/storage/provider/src/providers/rocksdb/invariants.rs

```diff
@@ -0,0 +1,870 @@
+    {
+        // Get the TransactionLookup stage checkpoint
+        let checkpoint = provider
+            .get_stage_checkpoint(StageId::TransactionLookup)?
+            .map(|cp| cp.block_number)
+            .unwrap_or(0);
+
+        // Find the max tx_number in RocksDB by iterating all entries.
+        // Note: We can't use `last()` because that returns the entry with the largest
+        // hash key, not the entry with the largest tx_number value.
+        let max_tx = self
+            .iter::<tables::TransactionHashNumbers>()?
+            .map(|r| r.map(|(_, tx_num)| tx_num))
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .max();
+
+        match max_tx {
+            Some(tx_number) => {
+                // If checkpoint is 0 but we have data, that's inconsistent
+                if checkpoint == 0 && tx_number > 0 {
+                    // RocksDB has data but stage hasn't run - need to clear all.
+                    // Prune all transactions in range [0, tx_number].
+                    self.prune_transaction_hash_numbers_in_range(provider, 0..=tx_number)?;
+                    return Ok(None);
+                }
+
+                // Map tx_number to block_number using TransactionBlocks table
+                // TransactionBlocks key is the highest tx ID in each block
+                let mut cursor = provider.tx_ref().cursor_read::<tables::TransactionBlocks>()?;
+
+                // Seek to the first entry where key >= tx_number
+                if let Some((_, rocks_block)) = cursor.seek(tx_number)? {
+                    // rocks_block is the block containing this transaction
+                    if checkpoint > rocks_block {
+                        // RocksDB is behind checkpoint - this indicates something went wrong
+                        // since MDBX checkpoint is committed last when appending.
+                        tracing::warn!(
+                            target: "reth::providers::rocksdb",
+                            rocks_max_tx = tx_number,
+                            rocks_block,
+                            checkpoint,
+                            "TransactionHashNumbers behind checkpoint, unwind needed"
+                        );
+                        return Ok(Some(rocks_block));
+                    } else if checkpoint < rocks_block {
+                        // RocksDB is ahead of checkpoint - this is the common case during
+                        // recovery from a crash during unwinding. MDBX checkpoint was
+                        // committed first, so we need to prune excess RocksDB data.
+                        //
+                        // Find the max tx_number for the checkpoint block by iterating
+                        // TransactionBlocks to find the entry where block_number == checkpoint
+                        let mut max_tx_for_checkpoint = 0u64;
+                        let mut cursor_walk =
```

Comment:
Given the order of storage commits, static files is at the very least at the same height as rocksdb. So we need to use that to find the range of transactions we need to pull.


* find out the last tx num in MDBX (can query for the last on `TransactionBlocks` or `BlockBodyIndices`)
* find out the highest tx num in static files (eg. `static_file_provider()
            .get_highest_static_file_tx(StaticFileSegment::Transactions))

`mdbx_last_tx_num+1..=highest_tx`